src/decision/src/LidarDecision.cpp

```diff
@@ -15,22 +16,115 @@ LidarDecision::LidarDecision(int argc, char **argv, std::string node_name) {
+    return twist_message_from_obstacle(most_dangerous_obstacle, max_obstacle_danger_distance,
+                                       obstacle_danger_angle, twist_velocity, twist_turn_rate);
+}
+
+std::vector<LidarObstacle> LidarDecision::findObstacles(const sensor_msgs::LaserScan &scan,
+                                                        float max_obstacle_angle_diff) {
+    // Get the raw scan data
+    std::vector<float> scan_data = scan.ranges;
+
+    // Find all ranges (lidar hits) that could be obstacles (initially each laser hit is an obstacle)
+    std::vector<LidarObstacle> obstacles;
+    for (int i = 0; i < scan_data.size(); i++) {
+        // Check that obstacle is within valid range
+        if (scan_data[i] < scan.range_max && scan_data[i] > scan.range_min) {
+            // If so, add it to the obstacles
+            obstacles.emplace_back(LidarObstacle(scan.angle_increment * i + scan.angle_min,
+                                                 scan_data[i]));
+        }
+    }
+
+    // Merge together obstacles that could be the same obstacle
+    mergeSimilarObstacles(obstacles, max_obstacle_angle_diff);
+
+    return obstacles;
+}
+
+LidarObstacle LidarDecision::mostDangerousObstacle(const std::vector<LidarObstacle> obstacles) {
+    // Return obstacle with the greatest danger score
+    return *std::max_element(obstacles.begin(), obstacles.end(),
+                              [&] (LidarObstacle obs1, LidarObstacle obs2){
+                                  return obs1.dangerScore() < obs2.dangerScore();
+                              });
+}
+
+void LidarDecision::mergeSimilarObstacles(std::vector<LidarObstacle>& obstacles, float max_angle_diff) {
+    // Ensure the list of obstacles is sorted in order of ascending angle
+    std::sort(obstacles.begin(), obstacles.end(),
+        [&] (LidarObstacle l1, LidarObstacle l2){
+            return l1.getAvgAngle() < l2.getAvgAngle();
+        });
+
+    // Merge similar obstacles
+    int i = 0;
+    while(i < obstacles.size() - 1){
+        // Check if angle difference between two consecutive scans is less then max_angle_diff
+        if (abs(obstacles[i+1].getMinAngle() - obstacles[i].getMaxAngle()) < max_angle_diff){
+            // Merge next obstacle into current one
+            obstacles[i].mergeInLidarObstacle(obstacles[i+1]);
+            obstacles.erase(obstacles.begin()+i+1);
+        } else {
+            i++;
+        }
+    }
+}
+
+geometry_msgs::Twist LidarDecision::twist_message_from_obstacle(LidarObstacle obstacle,
+                                                               distance_t danger_distance,
+                                                               angle_t danger_angle,
+                                                               float linear_vel_multiplier,
+                                                               float angular_vel_multiplier) {
+    geometry_msgs::Twist twist;
+    twist.linear.x = 0;
```

Comment:
I'm presuming that an all 0 message will be treated by the central decision node as no input (ie. it will either use other decision nodes or just go forward)